### ISSUE TITLE
Vip did not recover or relocate when manually deleted from master node

### ISSUE THREAD

**Issue by flerxu** [acassen/keepalived#534]
#Hello, everyone ! We have two server:

```
 web01     eth1: 192.168.16.101
 web02     eth1: 192.168.16.102
 virutal ip         : 192.168.16.100
```

As both keepalived daemons got up , vip was automatically attatched to web01. When either server was down or keepalived got crashed ,HA worked fine. But when I delete vip from web01 manually, vip did not recover on web01 or relocate to web02. 
```
# ip addr del 192.168.16.100/24 dev eth1
```

Keepalived configuration file on both servers:
```
[root@web02 ~]# more /etc/keepalived/keepalived.conf

vrrp_instance VI_1 {
    state MASTER
    interface eth1
    virtual_router_id 52
    priority 100
    advert_int 1
    authentication {
        auth_type PASS
        auth_pass 1111
    }
    virtual_ipaddress {
        192.168.16.100
    }
}
```

**Comment by pqarmitage** [comment]
First of all some comments on your configuration:
1. The valid states are MASTER and BACKUP. It is simplest to omit the `state` statement, but otherwise with priority < 255 it must be BACKUP
2. virtual_router_id must be a integer between 1 and 255
3. The virtual IP address should be /32, not /24. The subnet already exists, you are simply adding the address.

Now to the issue of manually deleting the VIP.
There is no code to detect or take any action in this event. The issue is what should happen. If you manually delete the VIP, we could assume you meant to do so, and therefore we should not do anything. On the other hand, that leaves the VRRP instance still advertising the address when it no longer has it.

This is further complicated by the ability to configure a VIP on a different interface. If that interface goes down, the address is removed by the kernel. What should keepalived do? It can't add the address back while the interface is down, and it would need to detect the interface coming back up in order to add it again.

And then there are also virtual_routes, and static_routes, which will be removed if they refer to an interface which goes down or an ip address which is deleted.

In fact, where an ip address or route specifies a device, then if it is desired that the vrrp instance should leave master state if the device goes down, the vrrp instance can be configured to track the interface (this still leaves a problem for static routes). But we don't have any mechanism where a route specifies an IP address that has been deleted.

My current thoughts are that if an ip address or route that a vrrp instance has added gets deleted, then the vrrp instance should transition out of master state. If the address and/or route can no longer be added due to a required device or address no longer being there, then the vrrp instance should enter fault state, but if it is simply that something like the address or route has been manually deleted, so that it could be added back again, then the instance should enter backup state, and then potentially transition back to master state thereby re-adding the address and/or route.

That leaves the issue of what to do about static routes. If it is considered that the route is needed for the virtual routers to operate, then all vrrp instances should transition to fault state. But it might be that different vrrp instances need different static routes. So this would probably need configuration against vrrp instances to identify which static routes they require.

Any thoughts on this from anyone would be most welcome.

**Comment by milari** [comment]
I'm not completely sure that's there isn't a bug in the code too.
Let me explain:
I've configured a check_script that simply returns output state:

#!/bin/bash
VIP_STATE=$(ssh $vIP date)
if [ "$?" -eq "0" ];then
 exit 0
else
 exit 1
fi

Now, when I do a ifdown interface (on MASTER node, that keeps vIP), nothing happens, as You explained.
But when I do ifup interface (later and always on MASTER node), nothing happens too...
In logs of both nodes, now I still see:

$(date) Keepalived_vrrp[7068]: /usr/local/bin/mycheckscript.sh exited with status 1

but if for BACKUP node is the same state of previous communication loss with MASTER node, after communication is restored, why MASTER doesn't pull up vIP again, if mycheckscript,sh returns exit 1?
Please, let me know.
Cheers.
M.

**Comment by pqarmitage** [comment]
I think the problem here is that once the interface goes down that vIP address is deleted and so the check_script starts returning 1, which means that the VRRP instance will be in fault state due to both the interface being down and the check script returning 1. When the interface comes back up, the IP address won't be added back because the vrrp instance is still in fault state, due to the check script returning 1, and the check script can't start returning 0 until the vrrp instance comes back up.

**Comment by milari** [comment]
but only 1 interface is really down (the Master one)!
so while Backup logging "1" state is correct, the Master one isn't (when if is up again).
Are You suggesting to delegate to an external script to keep back the vrrp istance when Master interface is back up?

**Comment by pqarmitage** [comment]
I think perhaps I haven't quite understood exactly what the issue is. Could you please describe again the exact sequence of events, and where and what the problem is, so that I can try and reproduce it.

**Comment by milari** [comment]
Ok.
- t0: node1 is MASTER, node2 is BACKUP; vIP is on Master
- t1: node1 is ifdown; node2 is BACKUP; no vIP
- t2 node1 is ifup; node2 is always BACKUP; still no vIP
- t3 both nodes report vrrp_istance is down and still no vIP
only way - now - to keep vIP up again is: 
- restart Master service keepalived
- ip addr add manually back
but in this second way (ip addr manually added), if a switch occurs for keepalived, vIP is on both nodes (Master and Backup), one by manually adding it, the other for switching vIP role.
Please let me know if I've not been clear enaugh.
:)

**Comment by sthakkel** [comment]
i have faced the same issue. no vIP is assigned to any nodes when both haproxy are working fine.

**Comment by pqarmitage** [comment]
Could you try the code from the beta branch and see how that behaves. It detects the deletion of VIPs and will make the vrrp instance transition to backup state.

**Comment by bahner** [comment]
Has this been committed to any releases since this issue, or has it only been commited to  the "beta" branch. I would try the beta-branch if that's the case.

**Comment by pqarmitage** [comment]
@bahner It has only been committed to the beta branch. I will not be independently merged into the master branch, but will be merge when the whole beta branch is merged into the master branch.

**Comment by acassen** [comment]
Beta branch is now close to be merged to master branch so I think we can close this issue.

**Comment by s33n** [comment]
vrrp_instance VI_1 {
    state MASTER
    interface p4p1
    virtual_router_id 51
    priority 100
    advert_int 1
    authentication {
        auth_type PASS
        auth_pass 1111
    }
    virtual_ipaddress {
        10.34.221.170
    }
}
virtual_server 10.34.221.170 80 {
    delay_loop 6
    lb_algo rr
    lb_kind NAT
    protocol TCP

    real_server 10.34.220.68 80 {
        TCP_CHECK {
                connect_timeout 10
        }
    }
    real_server 10.34.221.23 80 {
        TCP_CHECK {
                connect_timeout 10
        }

This is the code i had on my keepalived.conf. And It is switching from master to slave only incase of Keepalived service is down not with the cases where my realserver(master) is down.  Guys need your help correct me if im wrong i wanted to switch even in cases the Master IP is not reachable. 

Here are the var/log msgs

Jun 22 20:06:05 suri2 Keepalived_healthcheckers[7801]: TCP connection to [10.34.220.68]:80 failed !!!
Jun 22 20:06:05 suri2 Keepalived_healthcheckers[7801]: Removing service [10.34.220.68]:80 from VS [10.34.221.170]:80
Jun 22 20:06:05 suri2 Keepalived_healthcheckers[7801]: Remote SMTP server [0.0.0.0]:25 connected.
Jun 22 20:06:05 suri2 Keepalived_healthcheckers[7801]: SMTP alert successfully sent.
